src/Angor/Client/Services/MessageService.cs

```diff
@@ -0,0 +1,393 @@
+using Angor.Client.Models;
+using Angor.Shared.Services;
+using Angor.Shared.Utilities;
+using Microsoft.Extensions.Logging;
+using Nostr.Client.Client;
+using Nostr.Client.Keys;
+using Nostr.Client.Messages;
+using Nostr.Client.Messages.Direct;
+using Nostr.Client.Responses;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reactive.Linq;
+using System.Threading.Tasks;
+using Nostr.Client.Messages.Contacts;
+
+namespace Angor.Client.Services
+{
+    public class MessageService : IMessageService
+    {
+        private readonly ILogger<MessageService> _logger;
+        private readonly INostrCommunicationFactory _nostrCommunicationFactory;
+        private readonly IEncryptionService _encryptionService;
+        private readonly INetworkService _networkService;
+        private readonly IRelayService _relayService;
+        private readonly NostrConversionHelper _nostrHelper;
+
+        private List<DirectMessage> _directMessages = new();
+        private IDisposable _messageSubscription;
+        private string _currentUserPrivateKeyHex;
+        private string _currentUserNpub;
+        private string _currentUserHexPub;
+        private string _contactHexPub;
+        private bool _subscriptionActive = false;
+
+        public IReadOnlyList<DirectMessage> DirectMessages => _directMessages.AsReadOnly();
```

Comment:
`_directMessages` should probably be a dictionary where the eventid is the key, otherwise you will iterate many times to filter messages that dont need to be in the list